File: client/src/www/app/outline_server_repository/vpn.ts
// Copyright 2024 The Outline Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export const enum TunnelStatus {
  CONNECTED,
  DISCONNECTED,
  RECONNECTING,
  DISCONNECTING,
}

/** StartRequestJson is the serializable request to start the VPN, used for persistence and IPCs. */
export interface StartRequestJson {
  id: string;
  name: string;
  // Client config, ultimately passed to NewClient in Go.
  client: string;
  // First hop used by the legacy Electron code.
  firstHop: string;
}

/** VpnApi is how we talk to the platform-specific VPN API. */
export interface VpnApi {
  /**
   * Connects a VPN, routing all device traffic as described in the SessionConfig.
   * If there is another running instance, broadcasts a disconnect event and stops the active
   * tunnel. In such case, restarts tunneling while preserving the VPN.
   * @throws {OutlinePluginError}
   */
  start(request: StartRequestJson): Promise<void>;

  /** Stops the tunnel and VPN service. */
  stop(id: string): Promise<void>;

  /** Returns whether the tunnel instance is active. */
  isRunning(id: string): Promise<boolean>;

  /** Sets a listener, to be called when the tunnel status changes. */
  onStatusChange(listener: (id: string, status: TunnelStatus) => void): void;
}

File: client/go/outline/vpn.go
// Copyright 2024 The Outline Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package outline

import (
	"encoding/json"

	perrs "github.com/Jigsaw-Code/outline-apps/client/go/outline/platerrors"
	"github.com/Jigsaw-Code/outline-apps/client/go/outline/vpn"
)

type vpnConfigJSON struct {
	VPNConfig       vpn.Config `json:"vpn"`
	TransportConfig string     `json:"transport"`
}

// establishVPN establishes a VPN connection using the given configuration string.
// The configuration string should be a JSON object containing the VPN configuration
// and the transport configuration.
//
// The function returns a JSON string representing the established VPN connection,
// or an error if the connection fails.
func establishVPN(configStr string) (string, error) {
	var conf vpnConfigJSON
	if err := json.Unmarshal([]byte(configStr), &conf); err != nil {
		return "", perrs.PlatformError{
			Code:    perrs.IllegalConfig,
			Message: "invalid VPN config format",
			Cause:   perrs.ToPlatformError(err),
		}
	}

	// Create Outline Client and Device
	tcp, err := newFWMarkProtectedTCPDialer(conf.VPNConfig.ProtectionMark)
	if err != nil {
		return "", err
	}
	udp, err := newFWMarkProtectedUDPDialer(conf.VPNConfig.ProtectionMark)
	if err != nil {
		return "", err
	}
	c, err := newClientWithBaseDialers(conf.TransportConfig, tcp, udp)
	if err != nil {
		return "", err
	}
	proxy, err := NewDevice(c)
	if err != nil {
		return "", err
	}

	// Establish system VPN to the proxy
	conn, err := vpn.EstablishVPN(&conf.VPNConfig, proxy)
	if err != nil {
		return "", err
	}

	connJson, err := json.Marshal(conn)
	if err != nil {
		return "", perrs.PlatformError{
			Code:    perrs.InternalError,
			Message: "failed to return VPN connection as JSON",
			Cause:   perrs.ToPlatformError(err),
		}
	}
	return string(connJson), nil
}

// closeVPN closes the currently active VPN connection.
func closeVPN() error {
	return vpn.CloseVPN()
}
